File: web/packages/accounts/services/sessions.ts
import { authenticatedRequestHeaders, ensureOk } from "ente-base/http";
import { apiURL } from "ente-base/origins";
import { z } from "zod";

const Session = z.object({
    token: z.string(),
    isCurrent: z.boolean().optional(),
    creationTime: z.number(),
    ip: z.string(),
    ua: z.string(),
    prettyUA: z.string(),
    lastUsedTime: z.number(),
});

export type Session = z.infer<typeof Session>;

export const isCurrentSession = (
    session: Session,
    currentToken: string | undefined,
) => session.isCurrent ?? session.token === currentToken;

const SessionsResponse = z.object({ sessions: z.array(Session) });

export const getActiveSessions = async (): Promise<Session[]> => {
    const res = await fetch(await apiURL("/users/sessions"), {
        headers: await authenticatedRequestHeaders(),
    });
    ensureOk(res);
    const { sessions } = SessionsResponse.parse(await res.json());
    return sessions.sort((a, b) => b.lastUsedTime - a.lastUsedTime);
};

export const terminateSession = async (token: string): Promise<void> => {
    const url = await apiURL("/users/session");
    const res = await fetch(`${url}?token=${encodeURIComponent(token)}`, {
        method: "DELETE",
        headers: await authenticatedRequestHeaders(),
    });
    ensureOk(res);
};

File: mobile/packages/accounts/lib/models/sessions.dart
class Sessions {
  final List<Session> sessions;

  Sessions(this.sessions);

  factory Sessions.fromMap(Map<String, dynamic> map) {
    if (map["sessions"] == null) {
      throw Exception('\'map["sessions"]\' must not be null');
    }
    return Sessions(
      List<Session>.from(map['sessions']?.map((x) => Session.fromMap(x))),
    );
  }
}

class Session {
  final String token;
  final bool? isCurrent;
  final int creationTime;
  final String ip;
  final String ua;
  final String prettyUA;
  final int lastUsedTime;

  Session(
    this.token,
    this.isCurrent,
    this.creationTime,
    this.ip,
    this.ua,
    this.prettyUA,
    this.lastUsedTime,
  );

  factory Session.fromMap(Map<String, dynamic> map) {
    return Session(
      map['token'],
      map['isCurrent'],
      map['creationTime'],
      map['ip'],
      map['ua'],
      map['prettyUA'],
      map['lastUsedTime'],
    );
  }

  bool isCurrentSession(String? currentToken) =>
      isCurrent ?? token == currentToken;
}
